### PR TITLE
LOOP-3107 DIY Sync

### PR DESCRIPTION
**Pull request by ps2**
https://tidepool.atlassian.net/browse/LOOP-3107

### CODE REVIEW

File: Common/FeatureFlags.swift [hunk 1]
```diff
@@ -139,7 +155,8 @@ extension FeatureFlagConfiguration : CustomDebugStringConvertible {
             "* scenariosEnabled: \(scenariosEnabled)",
             "* sensitivityOverridesEnabled: \(sensitivityOverridesEnabled)",
             "* simulatedCoreDataEnabled: \(simulatedCoreDataEnabled)",
-            "* walshInsulinModelEnabled: \(walshInsulinModelEnabled)"
+            "* siriEnabled: \(siriEnabled)",
+            "* automaticBolusEnabled: \(automaticBolusEnabled)"
```
Comment (rickpasetto) [inline Common/FeatureFlags.swift]: add `outsideDosesEnabled`:
```suggestion
            "* automaticBolusEnabled: \(automaticBolusEnabled)",
            "* outsideDosesEnabled: \(outsideDosesEnabled)",
```

File: Loop.xcconfig [hunk 2]
```diff
@@ -44,3 +44,4 @@ LOOP_PROVISIONING_PROFILE_SPECIFIER_WATCHAPP_EXTENSION_RELEASE =
 
 // Optional overrides
 #include? "LoopOverride.xcconfig"
+#include? "../LoopConfigOverride.xcconfig"
```
Comment (rickpasetto) [inline Loop.xcconfig:47]: This looks wrong.  Why is this changed?

File: Loop/Extensions/DeviceDataManager+BolusEntryViewModelDelegate.swift [hunk 3]
```diff
@@ -10,8 +10,12 @@ import HealthKit
 import LoopCore
 import LoopKit
 
-extension DeviceDataManager: BolusEntryViewModelDelegate {
-
+extension DeviceDataManager: BolusEntryViewModelDelegate, LoggedDoseViewModelDelegate {
+    
+    func logOutsideInsulinDose(startDate: Date, units: Double, insulinType: InsulinType?) {
```
Comment (rickpasetto) [inline Loop/Extensions/DeviceDataManager+BolusEntryViewModelDelegate.swift]: Similar to the previous PR, I'm not a fan of the verb `log` when what it is actually doing is `store`.  Please consider changing all `logOutsideInsulinDose`s to `storeOutsideInsulinDose`.

File: Loop/View Models/LoggedDoseViewModel.swift [hunk 4]
```diff
@@ -0,0 +1,373 @@
...
+import LocalAuthentication
+import Intents
+import os.log
+import LoopCore
+import LoopKit
+import LoopKitUI
+import LoopUI
+import SwiftUI
+
+protocol LoggedDoseViewModelDelegate: AnyObject {
```
Comment (rickpasetto) [inline Loop/View Models/LoggedDoseViewModel.swift]: Consider `StoredDoseViewModel` and `StoredDoseViewModelDelegate`

File: Loop/View Models/LoggedDoseViewModel.swift [hunk 5]
```diff
@@ -0,0 +1,373 @@
...
+    
+    func withLoopState(do block: @escaping (LoopState) -> Void)
+
+    func logOutsideInsulinDose(startDate: Date, units: Double, insulinType: InsulinType?)
```
Comment (rickpasetto) [inline Loop/View Models/LoggedDoseViewModel.swift]: Consider `storeOutsideInsulinDose` instead.

File: Loop/Managers/DeviceDataManager.swift [hunk 6]
```diff
@@ -1169,13 +1183,48 @@ extension DeviceDataManager: LoopDataManagerDelegate {
...
 
-        log.default("LoopManager did recommend basal change")
+        log.default("LoopManager did recommend dose")
 
-        pumpManager.enactTempBasal(
-            unitsPerHour: basal.recommendation.unitsPerHour,
-            for: basal.recommendation.duration,
-            completion: completion
-        )
+        dosingQueue.async {
```
Comment (rickpasetto) [inline Loop/Managers/DeviceDataManager.swift]: Unit tests for this code change?

File: Loop/Info.plist [hunk 7]
```diff
@@ -61,8 +61,11 @@
 	<string>Meal data from the Health database is used to determine glucose effects. Glucose data from the Health database is used for graphing and momentum calculation. Sleep data from the Health database is used to optimize delivery of Apple Watch complication updates during the time you are awake.</string>
 	<key>NSHealthUpdateUsageDescription</key>
 	<string>Carbohydrate meal data entered in the app and on the watch is stored in the Health database. Glucose data retrieved from the CGM is stored securely in HealthKit.</string>
+	<key>NSSiriUsageDescription</key>
+	<string>Loop uses Siri to allow you to set override presets with your voice.</string>
```
Comment (rickpasetto) [inline Loop/Info.plist]: I think we no longer use the term "Override Preset".  Consider changing.

File: Loop/Managers/TrustedTimeChecker.swift [hunk 8]
```diff
@@ -33,14 +33,9 @@ class TrustedTimeChecker {
     private var ntpClient: TrueTimeClient
     private weak var alertManager: AlertManager?
     private lazy var log = DiagnosticLog(category: "TrustedTimeChecker")
-
+    
     init(alertManager: AlertManager) {
         ntpClient = TrueTimeClient.sharedInstance
-        #if DEBUG
```
Comment (rickpasetto) [inline Loop/Managers/TrustedTimeChecker.swift]: Why was this removed?  Seems like something we want to keep...

File: Loop/Models/LoopConstants.swift [hunk 9]
```diff
@@ -39,6 +39,8 @@ enum LoopConstants {
     // Compile time configuration
    
     static let retrospectiveCorrectionEnabled = true
+    
+    static let bolusPartialApplicationFactor = 0.4 // %
```
Comment (rickpasetto) [inline Loop/Models/LoopConstants.swift]: Please add code comment for what this represents.

File: Loop/View Models/BolusEntryViewModel.swift [hunk 10]
```diff
@@ -28,7 +28,9 @@ protocol BolusEntryViewModelDelegate: AnyObject {
 
     func storeBolusDosingDecision(_ bolusDosingDecision: BolusDosingDecision, withDate date: Date)
     
-    func enactBolus(units: Double, at startDate: Date, completion: @escaping (_ error: Error?) -> Void)
+    func enactBolus(units: Double, automatic: Bool, completion: @escaping (_ error: Error?) -> Void)
+    
+    func logOutsideInsulinDose(startDate: Date, units: Double, insulinType: InsulinType?)
```
Comment (rickpasetto) [inline Loop/View Models/BolusEntryViewModel.swift]: `logOutsideInsulinDose` -> `storeOutsideInsulinDose`

File: Loop/View Models/LoggedDoseViewModel.swift [hunk 11]
```diff
@@ -0,0 +1,373 @@
...
+                self?.delegate?.withLoopState { [weak self] state in
+                    self?.updatePredictedGlucoseValues(from: state)
+                }
+            }
+            .store(in: &cancellables)
+    }
+
+    // MARK: - View API
+
+    func logDose(onSuccess completion: @escaping () -> Void) {
```
Comment (rickpasetto) [inline Loop/View Models/LoggedDoseViewModel.swift]: `storeDose`

File: Common/FeatureFlags.swift [hunk 12]
```diff
@@ -114,10 +116,24 @@ struct FeatureFlagConfiguration: Decodable {
...
-        #if WALSH_INSULIN_MODEL_DISABLED
-        self.walshInsulinModelEnabled = false
+        #if SIRI_DISABLED
+        self.siriEnabled = false
         #else
-        self.walshInsulinModelEnabled = true
+        self.siriEnabled = true
+        #endif
+        
+        // Swift compiler config is inverse, since the default state is enabled.
```
Comment (rickpasetto) [inline Common/FeatureFlags.swift:125]: Shouldn't the default state be `disabled` for this?

File: Common/FeatureFlags.swift [hunk 13]
```diff
@@ -114,10 +116,24 @@ struct FeatureFlagConfiguration: Decodable {
         #endif
 
         // Swift compiler config is inverse, since the default state is enabled.
-        #if WALSH_INSULIN_MODEL_DISABLED
-        self.walshInsulinModelEnabled = false
+        #if SIRI_DISABLED
```
Comment (rickpasetto) [inline Common/FeatureFlags.swift:119]: Shouldn't the default state be `disabled` for this?

File: Common/FeatureFlags.swift [hunk 14]
```diff
@@ -114,10 +116,24 @@ struct FeatureFlagConfiguration: Decodable {
...
+        
+        // Swift compiler config is inverse, since the default state is enabled.
+        #if AUTOMATIC_BOLUS_DISABLED
+        self.automaticBolusEnabled = false
+        #else
+        self.automaticBolusEnabled = true
+        #endif
+        
+        // Swift compiler config is inverse, since the default state is enabled.
+        #if OUTSIDE_DOSES_DISABLED
```
Comment (rickpasetto) [inline Common/FeatureFlags.swift]: Shouldn't the default state be `disabled` for this?

File: Loop/Views/LoggedDoseView.swift [hunk 15]
```diff
@@ -0,0 +1,270 @@
...
+
+import Combine
+import HealthKit
+import SwiftUI
+import LoopKit
+import LoopKitUI
+import LoopUI
+
+
+struct LoggedDoseView: View {
```
Comment (rickpasetto) [inline Loop/Views/LoggedDoseView.swift]: consider `StoredDoseView`

File: Loop/View Controllers/InsulinDeliveryTableViewController.swift [hunk 16]
```diff
@@ -2,13 +2,651 @@
 //  InsulinDeliveryTableViewController.swift
 //  Naterade
 //
-//  Created by Nathan Racklyeft on 3/11/16.
+//  Created by Nathan Racklyeft on 1/30/16.
```
Comment (rickpasetto) [inline Loop/View Controllers/InsulinDeliveryTableViewController.swift:5]: I think this file was moved to `LoopKit`, so this diff probably shouldn't be here.

File: LoopTests/Extensions/DosingDecisionStoreTests.swift [hunk 17]
```diff
@@ -10,935 +10,936 @@ import XCTest
 import HealthKit
 
 @testable import Loop
-@testable import LoopKit
+// LoopKit is not compiled for Testing in carthage builds. LoopKit private units should be tested in LoopKit
+//@testable import LoopKit
```
Comment (rickpasetto) [inline LoopTests/Extensions/DosingDecisionStoreTests.swift]: This is a lot of unit tests being removed.  Were they moved to LoopKit, then?  If not, I have a concern over the reduction in unit test coverage.

File: LoopTests/Extensions/OutputStreamTests.swift [hunk 18]
```diff
@@ -8,40 +8,41 @@
 
 import XCTest
 import Foundation
-@testable import LoopKit
+// LoopKit is not compiled for Testing in carthage builds. LoopKit private units should be tested in LoopKit
```
Comment (rickpasetto) [inline LoopTests/Extensions/OutputStreamTests.swift:12]: Were these tests moved to LoopKit?

File: LoopTests/PersistenceControllerTestCase.swift [hunk 19]
```diff
@@ -8,38 +8,39 @@
...
-    func tearDown() {
-        managedObjectContext.performAndWait {
-            let coordinator = self.managedObjectContext.persistentStoreCoordinator!
-            let store = coordinator.persistentStores.first!
-            let url = coordinator.url(for: store)
-            try! self.managedObjectContext.persistentStoreCoordinator!.destroyPersistentStore(at: url, ofType: NSSQLiteStoreType, options: nil)
-        }
-    }
-}
+// LoopKit is not compiled for Testing in carthage builds. LoopKit private units should be tested in LoopKit
```
Comment (rickpasetto) [inline LoopTests/PersistenceControllerTestCase.swift]: Please consider an alternate solution to commenting this out.

File: LoopTests/ViewModels/LoggedDoseViewModelTests.swift [hunk 20]
```diff
@@ -0,0 +1,138 @@
...
+//  Copyright © 2021 LoopKit Authors. All rights reserved.
+//
+
+import HealthKit
+import LoopCore
+import LoopKit
+import XCTest
+@testable import Loop
+
+class LoggedDoseViewModelTests: XCTestCase {
```
Comment (rickpasetto) [inline LoopTests/ViewModels/LoggedDoseViewModelTests.swift]: `StoredDoseViewModelTests`

File: LoopUI/‏‏ar.lproj/Localizable.strings [hunk 21]
```diff
@@ -1,32 +0,0 @@
-/* Accessbility format value describing glucose: (1: glucose number)(2: glucose time) */
```
Comment (rickpasetto) [inline LoopUI/‏‏ar.lproj/Localizable.strings:1]: Why was this file deleted?

File: Scripts/carthage.sh [hunk 22]
```diff
@@ -16,4 +16,4 @@ echo 'EXCLUDED_ARCHS__EFFECTIVE_PLATFORM_SUFFIX_simulator__NATIVE_ARCH_64_BIT_x8
 echo 'EXCLUDED_ARCHS = $(inherited) $(EXCLUDED_ARCHS__EFFECTIVE_PLATFORM_SUFFIX_$(EFFECTIVE_PLATFORM_SUFFIX)__NATIVE_ARCH_64_BIT_$(NATIVE_ARCH_64_BIT)__XCODE_$(XCODE_VERSION_MAJOR))' >> $xcconfig
 
 export XCODE_XCCONFIG_FILE="$xcconfig"
-carthage "$@"
+"${SRCROOT}/bin/carthage" "$@"
```
Comment (rickpasetto) [inline Scripts/carthage.sh:19]: Please revert; we should not be checking in the `carthage` binary into SRCROOT.

File: Scripts/copy-frameworks.sh [hunk 23]
```diff
@@ -38,4 +38,4 @@ for COUNTER in $(seq 0 $(($SCRIPT_INPUT_FILE_COUNT - 1))); do
 done
 
 echo "Copy Frameworks with Carthage"
-carthage copy-frameworks
+"${SRCROOT}/bin/carthage" copy-frameworks
```
Comment (rickpasetto) [inline Scripts/copy-frameworks.sh:41]: Please revert; we should not be checking in the `carthage` binary into SRCROOT.

File: Version.xcconfig [hunk 24]
```diff
@@ -7,8 +7,8 @@
 //
 
 // Version [DEFAULT]
-LOOP_MARKETING_VERSION = 0.0.1
```
Comment (rickpasetto) [inline Version.xcconfig:10]: Please revert

File: Scripts/build-derived-assets.sh [hunk 25]
```diff
@@ -38,7 +38,14 @@ fi
 
 DERIVED_ASSETS="${DIRECTORY}/DerivedAssets.xcassets"
 DERIVED_ASSETS_BASE="${DIRECTORY}/DerivedAssetsBase.xcassets"
+
+# Assets can be overridden by a DerivedAssetsOverride.xcassets in ${DIRECTORY}, or
```
Comment (rickpasetto) [inline Scripts/build-derived-assets.sh:42]: This seems a little odd to me, but okay.  What is the purpose of it?

**Review by rickpasetto — CHANGES_REQUESTED** [review]
Some changes requested & some concerns over removal of unit tests.

Also, please consider removing `bin/carthage` (it shouldn't be checked in here)

File: Learn/Managers/DataManager.swift [hunk 26]
```diff
@@ -47,7 +47,8 @@ final class DataManager {
             healthStore: healthStore,
             cacheStore: cacheStore,
             observationEnabled: false,
-            insulinModel: insulinModelSettings?.model,
+            insulinModelProvider: PresetInsulinModelProvider(defaultRapidActingModel: defaultRapidActingModel),
+            longestEffectDuration: ExponentialInsulinModelPreset.rapidActingAdult.effectDuration,
```
Comment (nhamming) [inline Learn/Managers/DataManager.swift:51]: should this use the `defaultRapidActingModel`?

File: Loop Status Extension/StatusViewController.swift [hunk 27]
```diff
@@ -90,7 +90,8 @@ class StatusViewController: UIViewController, NCWidgetProviding {
         observeHealthKitSamplesFromOtherApps: FeatureFlags.observeHealthKitSamplesFromOtherApps,
         cacheStore: cacheStore,
         observationEnabled: false,
-        insulinModel: defaults?.insulinModelSettings?.model,
+        insulinModelProvider: PresetInsulinModelProvider(defaultRapidActingModel: defaults?.defaultRapidActingModel),
+        longestEffectDuration: ExponentialInsulinModelPreset.rapidActingAdult.effectDuration,
```
Comment (nhamming) [inline Loop Status Extension/StatusViewController.swift:94]: similar here about the `defaultRapidActingModel`.

File: Loop/Extensions/DosingDecisionStore+SimulatedCoreData.swift [hunk 28]
```diff
@@ -104,10 +104,12 @@ fileprivate extension StoredDosingDecision {
...
-        let recommendedTempBasal = StoredDosingDecision.TempBasalRecommendationWithDate(recommendation: TempBasalRecommendation(unitsPerHour: 0.75,
-                                                                                                                                duration: .minutes(30)),
-                                                                                        date: date.addingTimeInterval(-.minutes(1)))
-        let recommendedBolus = StoredDosingDecision.BolusRecommendationWithDate(recommendation: BolusRecommendation(amount: 0.2,
+        let tempBasalRecommendation = TempBasalRecommendation(unitsPerHour: 0.75,
+                                                              duration: .minutes(30))
+        let automaticDoseRecommendation = StoredDosingDecision.AutomaticDoseRecommendationWithDate(
+            recommendation: AutomaticDoseRecommendation(basalAdjustment: tempBasalRecommendation, bolusUnits: 0),
+            date: date.addingTimeInterval(-.minutes(1)))
+        let recommendedBolus = StoredDosingDecision.BolusRecommendationWithDate(recommendation: ManualBolusRecommendation(amount: 0.2,
```
Comment (nhamming) [inline Loop/Extensions/DosingDecisionStore+SimulatedCoreData.swift:112]: [nit] indent other parameters

File: Loop/Managers/DeviceDataManager.swift [hunk 29]
```diff
@@ -1151,12 +1162,15 @@ extension DeviceDataManager: LoopDataManagerDelegate {
             return units
         }
 
-        return pumpManager.roundToSupportedBolusVolume(units: units)
+        let rounded = ([0.0] + pumpManager.supportedBolusVolumes).enumerated().min( by: { abs($0.1 - units) < abs($1.1 - units) } )!.1
+        self.log.default("Rounded %{public}@ to %{public}@", String(describing: units), String(describing: rounded))
+
+        return rounded
```
Comment (nhamming) [inline Loop/Managers/DeviceDataManager.swift:1168]: I don't understand why the pumpManager isn't responsible for this rounding. Why do we want to round in the DeviceDataManager?

Comment (nhamming) [reply Loop/Managers/DeviceDataManager.swift:1168]: Also, if there is no pumpManager, why is `units` returned for a bolus that cannot be delivered? Could NaN (or similar) be used?

File: Loop/Managers/DoseMath.swift [hunk 30]
```diff
@@ -81,28 +83,47 @@ extension InsulinCorrection {
...
             pendingInsulin: pendingInsulin,
             notice: bolusRecommendationNotice
         )
     }
+    
+    /// Determines the bolus amount to perform a partial application correction
+    ///
+    /// - Parameters:
+    ///   - partialApplicationFactor: The fraction of needed insulin to deliver now
+    ///   - maxBolus: The maximum allowable bolus value in units
```
Comment (nhamming) [inline Loop/Managers/DoseMath.swift:114]: [nit] align parameter name

File: Loop/Managers/DoseMath.swift [hunk 31]
```diff
@@ -386,6 +406,88 @@ extension Collection where Element: GlucoseValue {
...
+    /// Returns nil if the normal scheduled basal, or active temporary basal, is sufficient.
+    ///
+    /// - Parameters:
+    ///   - correctionRange: The schedule of correction ranges
+    ///   - date: The date at which the temp basal would be scheduled, defaults to now
+    ///   - suspendThreshold: A glucose value causing a recommendation of no insulin if any prediction falls below
+    ///   - sensitivity: The schedule of insulin sensitivities
+    ///   - model: The insulin absorption model
+    ///   - basalRates: The schedule of basal rates
+    ///   - maxBasalRate: The maximum allowed basal rate
```
Comment (nhamming) [inline Loop/Managers/DoseMath.swift:421]: [nit] align parameter name

Comment (nhamming) [reply Loop/Managers/DoseMath.swift:421]: [nit] review this function description and update to align with the function.

File: Loop/Managers/ExtensionDataManager.swift [hunk 32]
```diff
@@ -140,7 +152,7 @@ final class StatusExtensionDataManager {
 }
 
 
-extension StatusExtensionDataManager: CustomDebugStringConvertible {
+extension ExtensionDataManager: CustomDebugStringConvertible {
```
Comment (nhamming) [inline Loop/Managers/ExtensionDataManager.swift:155]: [nit] I don't see the value in reducing the name to `ExtensionDataManager`. If this is needed, update line 158 accordingly.

File: Loop/Managers/LoopDataManager.swift [hunk 33]
```diff
@@ -1018,9 +1073,9 @@ extension LoopDataManager {
     ) throws -> [PredictedGlucoseValue] {
         dispatchPrecondition(condition: .onQueue(dataAccessQueue))
 
-        guard let model = insulinModelSettings?.model else {
-            throw LoopError.configurationError(.insulinModel)
-        }
+//        guard let insulinModelSettings = insulinModelSettings else {
+//            throw LoopError.configurationError(.insulinModel)
+//        }
```
Comment (nhamming) [inline Loop/Managers/LoopDataManager.swift]: [nit] remove?

File: Loop/Managers/LoopDataManager.swift [hunk 34]
```diff
@@ -1907,7 +1996,7 @@ protocol LoopDataManagerDelegate: AnyObject {
     ///
     /// - Parameters:
     ///   - units: The recommended bolus in U
-    /// - Returns: a supported bolus volume in U. The volume returned should not be larger than the passed in rate.
+    /// - Returns: a supported bolus volume in U. The volume returned should be the nearest deliverable volume.
```
Comment (nhamming) [inline Loop/Managers/LoopDataManager.swift:1991]: [nit] should this comment how rounding is handled?

File: Loop/Views/DosingStrategySelectionView.swift [hunk 35]
```diff
@@ -0,0 +1,70 @@
...
+    }
+}
+
+extension DosingStrategy {
+    var informationalText: String {
+        switch self {
+        case .tempBasalOnly:
+            return NSLocalizedString("Loop will set temporary basal rates to increase and decrease insulin delivery.", comment: "Description string for temp basal only dosing strategy")
+        case .automaticBolus:
+            return NSLocalizedString("Loop will automatically bolus when insulin needs are above scheduled basal, and will use temporary basal rates when needed to reduce insulin delivery below scheduled basal.", comment: "Description string for automatic bolus dosing strategy")
```
Comment (nhamming) [inline Loop/Views/DosingStrategySelectionView.swift:60]: why isn't `appName` being used here? Does this copy need review?

File: LoopCore/LoopSettings.swift [hunk 36]
```diff
@@ -8,6 +8,22 @@
 import HealthKit
 import LoopKit
 
+public enum DosingStrategy: Int, CaseIterable {
+    case tempBasalOnly
+    case automaticBolus
```
Comment (nhamming) [inline LoopCore/LoopSettings.swift:13]: [nit] doesn't `automaticBolus` mean `automaticDosing` (which includes temp basal and bolus)? automaticBolus reads like temp basal is not included.

File: LoopCore/LoopSettings.swift [hunk 37]
```diff
@@ -8,6 +8,22 @@
...
+}
+
+public extension DosingStrategy {
+    var title: String {
+        switch self {
+        case .tempBasalOnly:
+            return NSLocalizedString("Temp Basal Only", comment: "Title string for temp basal only dosing strategy")
+        case .automaticBolus:
+            return NSLocalizedString("Automatic Bolus", comment: "Title string for automatic bolus dosing strategy")
```
Comment (nhamming) [inline LoopCore/LoopSettings.swift:22]: [nit] should this be clarified that this automates both basal and bolus?

File: LoopTests/Extensions/OutputStreamTests.swift [hunk 38]
```diff
@@ -8,40 +8,41 @@
...
-        XCTAssertThrowsError(try outputStream.write(data)) { error in
-            XCTAssertEqual(error as! MockError, outputStream.error as! MockError)
-        }
-    }
-}
-
-fileprivate struct MockError: Error, Equatable {}
+//class OutputStreamTests: XCTestCase {
+//    private let string = "The quick brown fox jumps over the lazy dog"
+//
```
Comment (nhamming) [inline LoopTests/Extensions/OutputStreamTests.swift:17]: [nit] remove?

**Review by nhamming — COMMENTED** [review]
Personally, I would like to up the frequency of the DIY sync in hopes the resulting update to Tidepool Loop repos is reduces in number of changes. It is difficult to review such PRs well.

File: Loop.xcconfig (hunk 2 above)
Comment (ps2) [reply Loop.xcconfig:47]: Tidepool Loop and DIY Loop use different methods of overriding the config.  At some point Tidepool changed it from `../` to the local Loop directory (when the prepare workspace scripts were created), and it broke the DIY way of overriding. Adding back the the parent direction method with this.

File: Loop/Extensions/DeviceDataManager+BolusEntryViewModelDelegate.swift (hunk 3 above)
Comment (ps2) [reply Loop/Extensions/DeviceDataManager+BolusEntryViewModelDelegate.swift]: Yeah, this was all changed with the update to `manuallyEntered`

File: Loop/View Models/LoggedDoseViewModel.swift (hunk 4 above)
Comment (ps2) [reply Loop/View Models/LoggedDoseViewModel.swift]: changed with the update to manuallyEntered

File: Loop/View Models/LoggedDoseViewModel.swift (hunk 5 above)
Comment (ps2) [reply Loop/View Models/LoggedDoseViewModel.swift]: changed with the update to manuallyEntered

File: Loop/Managers/DeviceDataManager.swift (hunk 6 above)
Comment (ps2) [reply Loop/Managers/DeviceDataManager.swift]: Ok, adding some to cover this logic.

File: Loop/Info.plist (hunk 7 above)
Comment (ps2) [reply Loop/Info.plist]: Updated to "Loop uses Siri to allow you to enact presets with your voice."

File: Loop/Managers/TrustedTimeChecker.swift (hunk 8 above)
Comment (ps2) [reply Loop/Managers/TrustedTimeChecker.swift]: Yeah, I misunderstood what this was doing; reverted this change.

File: Loop/View Models/BolusEntryViewModel.swift (hunk 10 above)
Comment (ps2) [reply Loop/View Models/BolusEntryViewModel.swift]: changed with the update to manuallyEntered

File: Common/FeatureFlags.swift (hunk 12 above)
Comment (ps2) [reply Common/FeatureFlags.swift:125]: Why? This is a feature I'd assume we'd want in all builds, unless it needs to be disabled for some reason (i.e. Tidepool Loop).

File: Common/FeatureFlags.swift (hunk 14 above)
Comment (ps2) [reply Common/FeatureFlags.swift]: Not sure why we'd want default to be disabled.

File: Loop/Views/LoggedDoseView.swift (hunk 15 above)
Comment (ps2) [reply Loop/Views/LoggedDoseView.swift]: Changed.

File: Loop/View Controllers/InsulinDeliveryTableViewController.swift (hunk 16 above)
Comment (ps2) [reply Loop/View Controllers/InsulinDeliveryTableViewController.swift:5]: The original one was in LoopKit, but the new one is here; the old one hasn't been completely removed yet, but is named LegacyInsulinDeliveryTableViewController

File: LoopTests/Extensions/DosingDecisionStoreTests.swift (hunk 17 above)
Comment (ps2) [reply LoopTests/Extensions/DosingDecisionStoreTests.swift]: Yeah, the `@testable import LoopKit` should've never made it into Loop; I missed this in review, and will look into moving or redoing these tests so they don't depend on LoopKit internals.

File: LoopTests/Extensions/DosingDecisionStoreTests.swift [hunk 39]
```diff
@@ -8,340 +8,26 @@
 
 import XCTest
 import HealthKit
+import CoreData
+import LoopKit
 
 @testable import Loop
-@testable import LoopKit
```
Comment (ps2) [inline LoopTests/Extensions/DosingDecisionStoreTests.swift:13]: This test is redundant; functionality exercised here is tested in `testCodable` and `testExport`

File: Common/FeatureFlags.swift (hunk 13 above)
Comment (ps2) [reply Common/FeatureFlags.swift:119]: Why?

File: LoopTests/Extensions/DosingDecisionStoreTests.swift (hunk 17 above)
Comment (ps2) [reply LoopTests/Extensions/DosingDecisionStoreTests.swift]: Found a way to keep the tests and not depend on LoopKit internals (we can't use LoopKit internals in tests when building non-workspace ways, like Carthage or SPM); had to expose a cleanup method in LoopKit, which is in the associated LoopKit PR. One test was removed, but the functionality exercised in it was covered with other tests.

File: LoopTests/Extensions/OutputStreamTests.swift (hunk 18 above)
Comment (ps2) [reply LoopTests/Extensions/OutputStreamTests.swift:12]: See above.

File: LoopTests/PersistenceControllerTestCase.swift (hunk 19 above)
Comment (ps2) [reply LoopTests/PersistenceControllerTestCase.swift]: See above.

File: LoopTests/ViewModels/LoggedDoseViewModelTests.swift (hunk 20 above)
Comment (ps2) [reply LoopTests/ViewModels/LoggedDoseViewModelTests.swift]: Changed

File: LoopUI/‏‏ar.lproj/Localizable.strings (hunk 21 above)
Comment (ps2) [reply LoopUI/‏‏ar.lproj/Localizable.strings:1]: It's hard to see in github review here, but there are special characters in the filename. It actually breaks time machine.

File: Scripts/build-derived-assets.sh (hunk 25 above)
Comment (ps2) [reply Scripts/build-derived-assets.sh:42]: The original workspace builds overrode derived assets, for customizing Loop icons and such, by the existence of files in a parent directory (where the xcworkspace lives). Tidepool builds had to be updated to use a prepare script because there were two different xcworkspaces, and copy appropriate files in, but this parent directory method is still the build method used by DIY workspace builds.

File: Scripts/carthage.sh (hunk 22 above)
Comment (ps2) [reply Scripts/carthage.sh:19]: I agree that I generally wouldn't want to commit binaries here, but DIY builds need the carthage bin pinned; they will not build on a m1 mac with newer Carthage for now. It's a convenience I hope we can do away with soon, if the carthage issues are fixed, or we move to support SPM.

File: Version.xcconfig (hunk 24 above)
Comment (ps2) [reply Version.xcconfig:10]: DIY actually uses this for its version string. Tidepool uses the VersionOverride.

File: Common/FeatureFlags.swift (hunk 13 above)
Comment (rickpasetto) [reply Common/FeatureFlags.swift:119]: Because I would think that Product would want to weigh in on whether we want this in Tidepool Loop, and, if so, we should explicitly add it to the list of ENABLED features.  No?

File: Common/FeatureFlags.swift (hunk 12 above)
Comment (rickpasetto) [reply Common/FeatureFlags.swift:125]: Because I would think that Product would want to weigh in on whether we want this in Tidepool Loop, and, if so, we should explicitly add it to the list of ENABLED features.  No?

File: Loop/Managers/LoopDataManager.swift [hunk 40]
```diff
@@ -635,6 +661,38 @@ extension LoopDataManager {
...
+            if error == nil {
+                self.recommendedManualBolus = nil
+                self.recommendedAutomaticDose = nil
+                self.insulinEffect = nil
+                self.notify(forChange: .bolus)
+            }
+        }
+    }
+
+//    /// Logs a new external bolus insulin dose in the DoseStore and HealthKit
```
Comment (rickpasetto) [inline Loop/Managers/LoopDataManager.swift:685]: [nit] remove commented-out code?

File: LoopTests/Extensions/OutputStreamTests.swift [hunk 41]
```diff
@@ -8,40 +8,42 @@
 
 import XCTest
 import Foundation
-@testable import LoopKit
+import LoopKit
+// LoopKit is not compiled for Testing in carthage builds. LoopKit private units should be tested in LoopKit
+//@testable import LoopKit
```
Comment (rickpasetto) [inline LoopTests/Extensions/OutputStreamTests.swift:13]: [nit] Remove commented-out code?

File: Loop/Views/ManualEntryDoseView.swift [hunk 42]
```diff
@@ -0,0 +1,270 @@
...
+
+import Combine
+import HealthKit
+import SwiftUI
+import LoopKit
+import LoopKitUI
+import LoopUI
+
+
+struct ManualEntryDoseView: View {
```
Comment (rickpasetto) [inline Loop/Views/ManualEntryDoseView.swift:17]: [nit] Just a random thought here, but this shares a lot of similarity to `BolusEntryView` and `SimpleBolusView`.  Someday it would be nice to unify them...

**Review by rickpasetto — APPROVED** [review]
Some comments/nits to address, but other than that LGTM.

File: Learn/Managers/DataManager.swift (hunk 26 above)
Comment (ps2) [reply Learn/Managers/DataManager.swift:51]: No, they are independent; longest acting would be the longest of all available models, even if your default was a shorter one.

File: Loop Status Extension/StatusViewController.swift (hunk 27 above)
Comment (ps2) [reply Loop Status Extension/StatusViewController.swift:94]: same response

File: Loop/Managers/DeviceDataManager.swift (hunk 29 above)
Comment (ps2) [reply Loop/Managers/DeviceDataManager.swift:1168]: 1) Because how rounding is done is a Loop responsibility. Possible values are PumpManager responsibility, but if the dosing algorithm wants a value between possible values, Loop should decide which actual value is chosen.
2) Just a convenience to avoid having to deal with nils in display values, I'd guess, since this is also used for that. Could be changed if you feel strongly about it.

File: Loop/Managers/DoseMath.swift (hunk 30 above)
Comment (ps2) [reply Loop/Managers/DoseMath.swift:114]: I'm not sure I understand. The parameters are normally indented in this form, from what I see as the pattern.

File: Loop/Managers/ExtensionDataManager.swift (hunk 32 above)
Comment (ps2) [reply Loop/Managers/ExtensionDataManager.swift:155]: It now provides contexts for both the status extension, and the intents extension.

Comment (ps2) [reply Loop/Managers/ExtensionDataManager.swift:155]: Updating the other names, thanks.

File: Loop/Managers/LoopDataManager.swift (hunk 34 above)
Comment (ps2) [reply Loop/Managers/LoopDataManager.swift:1991]: Not 100% sure I understand what's being asked here. Are you suggesting that we should not comment on how it's implemented?

File: Loop/Views/DosingStrategySelectionView.swift (hunk 35 above)
Comment (ps2) [reply Loop/Views/DosingStrategySelectionView.swift:60]: Dosing strategy is not used in Tidepool Loop, so this isn't shown for now. If/when Tidepool wants to make this feature available, we'll need to review this and all the other aspects around automatic bolusing.

File: LoopCore/LoopSettings.swift (hunk 36 above)
Comment (ps2) [reply LoopCore/LoopSettings.swift:13]: This is the name of the dosing *strategy*. Even though this dosing strategy can modulate temp basals, by setting a rate less than scheduled, the users refer to this strategy as "automatic bolus". `automaticDosing` would apply to both strategies, as this setting specifies the strategy for automatic dosing.

File: LoopCore/LoopSettings.swift (hunk 37 above)
Comment (ps2) [reply LoopCore/LoopSettings.swift:22]: Again, this is just naming. I think you're right that users should understand this strategy can modulate temp basals (for reducing delivery), but that's more of a documentation/onboarding issue, I think.

File: LoopTests/Extensions/OutputStreamTests.swift (hunk 38 above)
Comment (ps2) [reply LoopTests/Extensions/OutputStreamTests.swift:17]: Fixed the LoopKit import, and uncommented.

File: Common/FeatureFlags.swift (hunk 13 above)
Comment (ps2) [reply Common/FeatureFlags.swift:119]: Siri authorization is currently only needed for activating presets via voice. I had this disabled because Tidepool Loop is not supporting DIY style override presets yet, but now that I think about it, perhaps we can update this to activate the old Pre-Meal and Activity presets. I'll create a Jira ticket for that.

File: Common/FeatureFlags.swift (hunk 14 above)
Comment (ps2) [reply Common/FeatureFlags.swift]: See above on preset activation.
